aten/src/ATen/native/sparse/SparseTensorMath.cpp

```diff
@@ -928,6 +935,31 @@ SparseTensor& mul_out_sparse_cpu(const Tensor& t_, const Tensor& src_, Tensor& r
     return _mul_dense_sparse_out(t_, src_, r);
   }
 
+  // case mul(sparse, sparse) with a 0-dim input.
+  const auto get_vals_from_scalar = [](const SparseTensor& s) -> Tensor {
+    const auto vals = s._values().squeeze(0);
+    if (vals.dim()) {
+      return at::empty({}, vals.options());
+    }
+    return vals;
+  };
+  const auto probably_scalar = (!src_.dim() && src_.is_coalesced())
+    ? src_
+    : (!t_.dim() && t_.is_coalesced())
+    ? t_ : Tensor {};
+  const auto other = is_same_tensor(probably_scalar, t_) ? src_ : t_;
+  // Case when there is a 0-dim argument which is coalesced.
+  if (probably_scalar.defined()) {
+    return _mul_dense_sparse_out(get_vals_from_scalar(probably_scalar), other, r);
+  }
+  // case for 0-dim arguments that need to get coalesced
+  if (!src_.dim()) {
+    return _mul_dense_sparse_out(get_vals_from_scalar(src_.coalesce()), t_, r);
+  }
+  if (!t_.dim()) {
+    return _mul_dense_sparse_out(get_vals_from_scalar(t_.coalesce()), src_, r);
+  }
```

Comment:
re lines 946-961 - I totally get the desire to avoid big branchy if/elseif nests, but the logic here seems kind of complicated and I think there's some avoidable redundancy. 

WDYT of something like this? Assuming I've read the original correctly, it does the same thing and avoids the double `!dim()` testing and use of a fresh Tensor as a "none" value:
```
if (!src_.dim()) {
  vals = get_vals_from_scalar(src_.is_coalesced() ? src_ : src_.coalesce());
  return _mul_dense_sparse_out(vals, t_, r);
}
if (!t_.dim()) {
  vals = get_vals_from_scalar(t_.is_coalesced() ? t_ : t_.coalesce());
  return _mul_dense_sparse_out(vals, src_, r);
}
```

Reply:
Yeah, I like the suggestion, but I wanted to postpone the call to coalesce as much as possible, as it is expensive (sync plus synchronizations).

Reply:
Shouldn't coalesce short circuit to a no-op if `is_coalesced()`?

Reply:
From the code suggested it could be that src is not coalesced and t is coalesced, and we end up calling coalesce when it could be avoided.

Reply:
I think the simplification would call coalesce in the same places/at the same time as the original, no? Maybe I'm missing it - sorry to make you spell it out, but can you give me an example where the coalesce behavior would be different?

Reply:
Imagine that src is non-coalesced, t is coalesced, and both of them are 0-dim.

Reply:
Ah got it, thanks! And yeah, agree it's probably worth using an optional rather than the Tensor {} sentinel.